lightning/src/util/persist.rs

```diff
@@ -205,3 +213,760 @@ where
 	}
 	Ok(res)
 }
+
+struct MonitorUpdatingPersister<K: Deref, L: Deref>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	kv_store: K,
+	logger: L,
+}
+
+#[allow(dead_code)]
+impl<K: Deref, L: Deref> MonitorUpdatingPersister<K, L>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	pub fn read_channel_monitors_with_updates<
+		ES: Deref + Clone,
+		SP: Deref + Clone,
+		B: Deref,
+		F: Deref + Clone,
+	>(
+		&self,
+		entropy_source: ES,
+		signer_provider: SP,
+		broadcaster: &B,
+		fee_estimator: F,
+	) -> io::Result<Vec<(BlockHash,ChannelMonitor<<SP::Target as SignerProvider>::Signer>)>>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+		B::Target: BroadcasterInterface,
+		F::Target: FeeEstimator,
+	{
+		let mut res = Vec::new();
+		// for each monitor...
+		for monitor_name in self.list_monitor_names()? {
+			// ...parse the monitor
+			let (bh, monitor) = self.deserialize_monitor(
+				entropy_source.clone(),
+				signer_provider.clone(),
+				&monitor_name,
+			)?;
+			// ...parse and apply the updates with an id higher than the monitor.
+			for update_name in self.list_update_names(&monitor_name)? {
+				let update = self.deserialize_monitor_update(&monitor_name, &update_name)?;
+				if (update.update_id == CLOSED_CHANNEL_UPDATE_ID
+					|| update.update_id > monitor.get_latest_update_id())
```

Comment:
We _could_. 

Initially this is solely how I did GC. I didn't like that because it means your storage grows with uptime, and you also add more work to the init. If you spread the GC work out, generally it should be less noticeable and better tolerated by storage, as well as keeping growth in check. And once I had that, I didn't want to have two different ways that updates get GC'd. I also think LDK rewrites all the monitors at init, which would prompt the cleanup pass anyway.

However, at this point it's probably pretty easy to send a single delete to `KVStore` here, and I don't think it will be too hard to reason about.

Reply:
Hm, on second thought -- is there not value (to clarity, and perhaps surprise-avoidance) in having this method be non-mutative to storage?

Reply:
I think this is also outdated after discussion on the MUP RFC.